File: nvpl_blas/c/ztrsv.c
/******************************************************************************
 * Content: 
 *     This example demonstrates use of API as below:
 *     cblas_ztrsv
 *
 ******************************************************************************/
#include "example_helper.h"

int main() {
    nvpl_int_t N = 2;
    nvpl_int_t incX = 1;
    nvpl_int_t lda = N;
    enum CBLAS_ORDER order = CblasRowMajor;
    enum CBLAS_UPLO uplo = CblasUpper;
    enum CBLAS_TRANSPOSE trans = CblasNoTrans;
    enum CBLAS_DIAG diag = CblasNonUnit;
    nvpl_dcomplex_t * A = NULL;
    nvpl_dcomplex_t * X = NULL;
    nvpl_int_t len_x = 1 + (N - 1) * labs(incX);

    printf("\nExample: cblas_ztrsv for solving a system of linear equations whose coefficients are in a triangular matrix\n\n");
    printf("#### args: n=%" PRId64 ", lda=%" PRId64 ", incX=%" PRId64 ", order=%c, uplo=%c, trans=%c, diag=%c\n",
           (int64_t)N, (int64_t)lda, (int64_t)incX, order_to_char(order), uplo_to_char(uplo), transpose_to_char(trans),
           diag_to_char(diag));

    // allocate memory
    A = (nvpl_dcomplex_t *)malloc(lda * N * sizeof(nvpl_dcomplex_t));
    X = (nvpl_dcomplex_t *)malloc(len_x * sizeof(nvpl_dcomplex_t));

    // fill data
    fill_zmatrix(A, N, N, lda, order, UPLO_TO_FILL_MODE(uplo), diag);
    fill_zvector(X, N, incX);

    // print input data
    print_zmatrix(A, N, N, lda, "A", order);
    print_zvector(X, N, incX, "X");

    // call cblas_ztrsv
    cblas_ztrsv(order, uplo, trans, diag, N, A, lda, X, incX);

    // print result
    printf("\nAfter ztrsv operation:\n");
    print_zvector(X, N, incX, "X");

    // release memory
    free(A);
    free(X);
    return 0;
}

File: nvpl_blas/f77/ztrsv.f
!******************************************************************************
! Content:
!     This example demonstrates use of API as below:
!     ZTRSV
!
!******************************************************************************
      program   ZTRSV_MAIN
*
      implicit none

      include "nvpl_blas_f77_blas.fi"

*       Intrinsic Functions
      intrinsic  abs

      integer         n, incx, i, lda
      parameter       (n=2, incx=1, lda=2)
      integer         xsize, apsize
      parameter       (xsize=2+(n-1)*abs(incx))
      double complex  x(xsize), a(lda,n)

*       External Subroutines
      external        print_zvector, fill_zvector,
     &                print_zmatrix, fill_zmatrix
      character*1     uplo, trans, diag
      parameter       (uplo='U', trans='N', diag='N')
*
*      Executable Statements
*
      print*
      print 99
*
*       Print input data
      print*
      print 100, n, lda, uplo, trans, diag, incx
      call fill_zmatrix(a,n,n,lda,uplo,diag)
      call fill_zvector(x,n,incx)
      call print_zvector(x,n,incx,'X')
      call print_zmatrix(a,n,n,lda,'A')
*
*      Call ZTRSV subroutine
      call ZTRSV(uplo,trans,diag,n,a,lda,x,incx)
*
      print*
      print 102
      call print_zvector(x,n,incx,'X')

      stop
 99   format('Example: ZTRSV for solving a system of linear equations ',
     &       'whose coefficients are in a triangular matrix')
 100  format('#### args: n=',i1,', lda=',i1,', uplo=',a1,', trans=',a1,
     &       ', diag=',a1,', incx=',i1)
 102  format('The result of ZTRSV: ')
      end
